src/Motor.Extensions.Hosting.CloudEvents/MotorCloudEvent.cs

```diff
@@ -0,0 +1,114 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Mime;
+using CloudNative.CloudEvents;
+
+namespace Motor.Extensions.Hosting.CloudEvents
+{
+    public static class MotorCloudEventInfo
+    {
+        public static CloudEventsSpecVersion SpecVersion => CloudEventsSpecVersion.V1_0;
+
+        public static readonly IEnumerable<CloudEventAttribute> RequiredAttributes = SpecVersion.RequiredAttributes.Concat(new[]{
+            SpecVersion.TimeAttribute
+        });
+    }
+
+    public class MotorCloudEvent<TData> where TData : class
+    {
+        private readonly IApplicationNameService _applicationNameService;
+
+        private MotorCloudEvent(
+            IApplicationNameService applicationNameService,
+            TData data,
+            Uri source,
+            string? id = null,
+            DateTime? time = null,
+            params KeyValuePair<CloudEventAttribute, object>[] extensions)
+            : this(applicationNameService, data, typeof(TData).Name, source, id, time, extensions)
+        {
+        }
+
+        public MotorCloudEvent(
+            IApplicationNameService applicationNameService,
+            TData data,
+            string type,
+            Uri source,
+            string? id = null,
+            DateTimeOffset? time = null,
+            params KeyValuePair<CloudEventAttribute, object>[] extensions)
+            : this(applicationNameService, data, extensions)
```

Comment:
Now that you say it, it looks like the only difference between both constructors is that one expects the `type` to be set explicitly while the other one infers it from `data`. I think it would be a good idea to merge them.

How would you call the constructor without having `BaseEvent.Time` set? `time` defaults to `null`, but in that case, `BaseEvent.Time` will be set to `DateTimeOffset.UtcNow`.

Reply:
> How would you call the constructor without having BaseEvent.Time set?

I meant the third constructor here which doesn't have a `time` parameter. Is it valid have a cloud event without having set `Type`, `Source`, `Time`, and `DataContentType`? Because if that's the case, then it's also OK to have the constructors like that.
It just seemed to me like those properties should be set.

But if that's not the case, then I would at least suggest to only assign `Id` in the third constructor and add `string? id = null` as an optional argument to it. That way, the `Id` property will at least not be assigned twice.

Reply:
Except `DataContentType`, all of these attributes are mandatory. I merged all three constructors into one.